packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -124,62 +178,88 @@ export const useGridVirtualScroller = () => {
-      prevTotalWidth.current = dimensions.columnsTotalWidth;
+      previousContextScrollPosition.current = scrollPosition.current;
     },
-    [apiRef, dimensions.isReady, dimensions.columnsTotalWidth],
+    [apiRef, dimensions.isReady],
   );
 
   const triggerUpdateRenderContext = () => {
-    const inputs = inputsSelector(apiRef, rootProps, enabled, enabledForColumns);
-    const [nextRenderContext, rawRenderContext] = computeRenderContext(inputs, scrollPosition);
+    const newScroll = {
+      top: scrollerRef.current!.scrollTop,
+      left: scrollerRef.current!.scrollLeft,
+    };
 
-    // Since previous render, we have scrolled...
-    const topRowsScrolled = Math.abs(
-      rawRenderContext.firstRowIndex - previousContext.current.firstRowIndex,
-    );
-    const bottomRowsScrolled = Math.abs(
-      rawRenderContext.lastRowIndex - previousContext.current.lastRowIndex,
-    );
+    const dx = newScroll.left - scrollPosition.current.left;
+    const dy = newScroll.top - scrollPosition.current.top;
+
+    const isScrolling = dx !== 0 || dy !== 0;
+
+    scrollPosition.current = newScroll;
 
-    const topColumnsScrolled = Math.abs(
-      rawRenderContext.firstColumnIndex - previousContext.current.firstColumnIndex,
+    const direction = isScrolling ? directionForDelta(dx, dy) : ScrollDirection.NONE;
+
+    // Since previous render, we have scrolled...
+    const rowScroll = Math.abs(
+      scrollPosition.current.top - previousContextScrollPosition.current.top,
     );
-    const bottomColumnsScrolled = Math.abs(
-      rawRenderContext.lastColumnIndex - previousContext.current.lastColumnIndex,
+    const columnScroll = Math.abs(
+      scrollPosition.current.left - previousContextScrollPosition.current.left,
     );
 
-    const shouldUpdate =
-      topRowsScrolled >= rootProps.rowThreshold ||
-      bottomRowsScrolled >= rootProps.rowThreshold ||
-      topColumnsScrolled >= rootProps.columnThreshold ||
-      bottomColumnsScrolled >= rootProps.columnThreshold ||
-      prevTotalWidth.current !== dimensions.columnsTotalWidth;
+    const didCrossThreshold = rowScroll >= dimensions.rowHeight || columnScroll >= 50;
+    const didChangeDirection = scrollCache.direction !== direction;
+    const shouldUpdate = didCrossThreshold || didChangeDirection;
 
     if (!shouldUpdate) {
-      return previousContext.current;
+      return renderContext;
+    }
+
+    // Render a new context
+
+    if (didChangeDirection) {
+      switch (direction) {
+        case ScrollDirection.NONE:
+        case ScrollDirection.LEFT:
+        case ScrollDirection.RIGHT:
+          frozenContext.current = undefined;
+          break;
+        default:
+          frozenContext.current = renderContext;
+          break;
+      }
     }
 
+    scrollCache.direction = direction;
+    scrollCache.buffer = bufferForDirection(
+      direction,
+      rootProps.rowBufferPx,
+      rootProps.columnBufferPx,
+      dimensions.rowHeight * 15,
+      50 * 3,
```

Comment:
Why is the horizontal buffer much smaller than the vertical one?
Assuming the default row height, it's 780px vs only 150px.

Reply:
Could be increased, but those numbers felt good in practice. If you have a different proposition I'm ok with trying it out.

Reply:
How about 300px?